### ISSUE TITLE
State of selection of "deleteContentBackward" and "deleteContentForward" is broken and is not consistent with the other inputType values

### ISSUE THREAD

**Issue by masayuki-nakano** [w3c/input-events#82]
> `deleteContentBackward`
> delete the content directly before the caret position and this intention is not covered by another                            inputType or delete the selection with the selection collapsing to its start after the deletion
> State of selection: Non-collapsed

> `deleteContentForward`
> delete the content directly after the caret position and this intention is not covered by another                            inputType or delete the selection with the selection collapsing to its end after the deletion
> State of selection: Non-collapsed

Those inputType values are defined as those explanations.  However,

> `deleteWordBackward`
> delete a word directly before  the caret position
> State of selection: Collapsed

is defined like this and other delete*Backward and delete*Forward are defined similarly. Then, there is:

> `deleteContent`
> delete the selection without specifying the direction of the deletion and this intention is not covered by another inputType
> State of selection: Non-collapsed

Safari 12 behaves exactly same as defined here (Chrome is broken around here, see <https://bugs.chromium.org/p/chromium/issues/detail?id=899624>). However, from point of view of web apps, only `deleteContentBackward` and `deleteContentForward` sound like fired only when selection is collapsed and when there is non-collapsed selection, `deleteContent` should be fired.

Anyway, `Non-collapsed` in the first 2 declarations are wrong. Should be `Any` at least. But perhaps, better state is `Collapsed`.

ccing @rniwa -san.

**Comment by masayuki-nakano** [comment]
ccing @johanneswilm

**Comment by johanneswilm** [comment]
Hey @masayuki-nakano 

> However, from point of view of web apps, only deleteContentBackward and deleteContentForward sound like fired only when selection is collapsed and when there is non-collapsed selection, deleteContent should be fired.

Actually it does make a difference. Sure, in the case that the content is actually deleted, the caret will be placed at the same place after the deletion has taken place (the deletion point). But JS editors may want to treat the two actions differently, for example in a tracked changes context: The content is then not actually deleted, but just marked as deleted and the caret is collapsed to one of the sides. Which side it is collapsed to depends on the direction of the deletion. So that is why you need to use those two different inputTypes.

Does that make sense or should I add drawings/descriptions?

**Comment by johanneswilm** [comment]
I agree though that the state of the selection should be "Any".

**Comment by masayuki-nakano** [comment]
> But JS editors may want to treat the two actions differently, for example in a tracked changes context: The content is then not actually deleted, but just marked as deleted and the caret is collapsed to one of the sides.

When I tested with Chrome, I thought so too. However, if it's true, why do, for example, `deleteWordBackward`, should be set only when Selection is collapsed?

Therefore, I think that if `deleteContentBackward` and `deleteContentForward` should be declared as `Any`, probably, `deleteWordBackward`, `deleteWordForward`, `deleteSoftLineBackward`, `deleteSoftLineForward`, `deleteEntireSoftLine`, `deleteHardLineBackward` and `deleteHardLineForward` should also be `Any`. (Perhaps, this is better for web apps.)

**Comment by masayuki-nakano** [comment]
(If so, Chrome should keep their behavior but Safari needs to change their behavior)

**Comment by johanneswilm** [comment]
@masayuki-nakano I fixed the wording the part that was obviously wrong in the above commit and what this issue is about. I consider it to be on the level of a typo:

> Therefore, I think that if deleteContentBackward and deleteContentForward should be declared as Any,

> deleteWordBackward, deleteWordForward, deleteSoftLineBackward, deleteSoftLineForward, deleteEntireSoftLine, deleteHardLineBackward and deleteHardLineForward should also be Any. (Perhaps, this is better for web apps.)

Not the way we have been using those so far. If the user has content selected, no matter what that content it, it should use `deleteContentBackward`, `deleteContentForward` if the direction is known or `deleteContent` if the direction is not know. In these cases, it does not matter what content is selected (a word, a few letters, an entire paragraph, etc.). What is to be deleted is equal to the selection, no matter what that is.

It is only if your UA has a special command to delete something other than the selection that we use the other inputTypes.

So for example, take this situation:

> This is an |important sentence.
> | = caret

And the user hits the keycode-combination for "delete word forward". In this case we should receive the beforeinput event with an inputType of `deleteWordForward`. 

However, if instead we instead have this situation:

> This is an [important] sentence.
> [] = selection

And the user hits the backspace key. Now the beforeinput should have an inputType of `deleteContentBackward`. This is why we only use all those other deletion types with a collapsed selection. Does that make sense?

I will close this issue as the original issue has been resolved. If you are not satisfied by my answer, please copy your comment into a new issue because then we probably need to discuss it some more.

**Comment by masayuki-nakano** [comment]
> Not the way we have been using those so far. If the user has content selected, no matter what that content it, it should use `deleteContentBackward`, `deleteContentForward` if the direction is known or `deleteContent` if the direction is not know.

Well, do you mean that if Selection is not collapsed but user presses key combination to do `deleteWordBackward`, `deleteWordForward`, `deleteSoftLineBackward`, `deleteSoftLineForward`, `deleteEntireSoftLine`, `deleteHardLineBackward` or `deleteHardLineForward`, browsers should set `deleteContentBackward` or `deleteContentForward` rather than `deleteContent`? If so, it should be explicitly declared by the spec and Chrome and Safari should change their behavior.

Note that due to the difference of market share, Firefox cannot take other behavior than Chrome and Safari without special reasons. So, I really need how Spec fixes such inconsistent issue, and to know signals of each browser vender which has already implemented something new.

**Comment by johanneswilm** [comment]
@masayuki-nakano My understanding is that the keycodes to delete words forward/backward, etc. are only available to the end user if the selection is collapsed. If the selection is not collapsed, it simply deletes whatever is in the selection. Or it first collapses the selection and then does the deletion as a second step.

Have you come across a UA that does all for those actions even though the selection is not collapsed?

So say you have this:

> The dog is old and[ gre]y.
> [] = selection

And the user now hits the key code for 'delete word backward' and so now the word "and" is deleted? while " gre" stays as it is and the selection also stays there? Like this:

> The dog is old [ gre]y.

If this is the case on some UA, then yes, we need to change the spec. If this is not possible though, or the selection is first collapsed before the 'delete word backward' is executed, then I believe the spec should stay as it is.

**Comment by johanneswilm** [comment]
@masayuki-nakano I just tried this on a Mac/Safari in contenteditable:

1. Select some text in a paragraph.
2. Hit Control+k
3. Nothing happens
4. Collapse the selection.
5. Hit Control+k again
6. Notice how everything from the collapsed selection to the end of the hardlinebreak is deleted.

So this is the case where we use that inputType. The key binding is simply deactivated while the selection is not collapsed, which is why the the spec text so far is correct.

However, if you now tell me that Firefox works differently and that in Firefox the key binding is active also when content is selected, then we yes we do need to add this to the spec. 

With the spec we simply want to list all the pre-eisting editing operations provided by at least one browser. We don't try to cover any kind of editing operation anyone could think of some time in the future.

Does that make sense?

**Comment by masayuki-nakano** [comment]
@johanneswilm 

> My understanding is that the keycodes to delete words forward/backward, etc. are only available to the end user if the selection is collapsed. If the selection is not collapsed, it simply deletes whatever is in the selection. Or it first collapses the selection and then does the deletion as a second step.

Yes, me too.

> Have you come across a UA that does all for those actions even though the selection is not collapsed?

No, I'm talking about inconsistency of current spec. Current spec declares:

| ___________________________________ | Selection is collapsed__ | selection is NOT collapsed |
| `Backspace` only__________________ | `deleteContentBackward` | `deleteContentBackward`___ |
| `Ctrl` + `Backspace` (assuming to |
| delete to begin of word)_________ | `deleteWordBackward`____ | `deleteContent`___________ |

However, for consistency, shouldn't `deleteContent` in this table be `deleteContentBackward`? If you believe that web apps do not need the direction information, why do web apps need to know it only when `Backspace` or `Delete` without modifiers simply pressed?  So, I think that spec should declare as:

| ___________________________________ | Selection is collapsed__ | selection is NOT collapsed |
| `Backspace` only__________________ | `deleteContentBackward` | `deleteContent`___________ |
| `Ctrl` + `Backspace` (assuming to |
| delete to begin of word)_________ | `deleteWordBackward`____ | `deleteContent`___________ |

or

| ___________________________________ | Selection is collapsed__ | selection is NOT collapsed |
| `Backspace` only__________________ | `deleteContentBackward` | `deleteContentBackward`___ |
| `Ctrl` + `Backspace` (assuming to |
| delete to begin of word)_________ | `deleteWordBackward`____ | `deleteContentBackward`___ |

>>  The dog is old and[ gre]y.
>>  [] = selection
> 
> And the user now hits the key code for 'delete word backward' and so now the word "and" is deleted? while " gre" stays as it is and the selection also stays there? Like this:
>  
>> The dog is old [ gre]y.
> 
> If this is the case on some UA, then yes, we need to change the spec. If this is not possible though, or the selection is first collapsed before the 'delete word backward' is executed, then I believe the spec should stay as it is.

I think that all browsers remove selected content in this case. So, if this should be `deleteContent`, `Backspace` key without modifier with non-collapsed selection should also be `deleteContent` rather than `deleteContentBackward`.

> However, if you now tell me that Firefox works differently and that in Firefox the key binding is active also when content is selected, then we yes we do need to add this to the spec.

No, that's not my point. My point is, the inconsistency.

> With the spec we simply want to list all the pre-eisting editing operations provided by at least one browser. We don't try to cover any kind of editing operation anyone could think of some time in the future.
> 
> Does that make sense?

Yes, I agree with that. I'm trying to say that I don't understand the inconsistency. You said:

> But JS editors may want to treat the two actions differently, for example in a tracked changes context: The content is then not actually deleted, but just marked as deleted and the caret is collapsed to one of the sides. Which side it is collapsed to depends on the direction of the deletion. So that is why you need to use those two different inputTypes.

So, `Backward` or `Forward` is important information in this idea. However, current spec sounds like that it's not important only when user tries to remove content to word or visual line boundary. So, my idea is, for example, `deleteWordBackward` should be declared as "delete a word directly before the caret position. **If selection is not collapsed and browser removes the selected content instead, set `deleteContentBackward`.**".

**Comment by johanneswilm** [comment]
> However, for consistency, shouldn't deleteContent in this table be deleteContentBackward? 

I agree, but that's what is declared in the spec now, right? `deleteContentForward` and `deleteContentBackward` should be used for `Delete` and `Backspace`. I think the wording is quite clear on that.

The `deleteContent` should only be used when there no direction is specified. So this would not be with any of the usual keys, but a UA may for example have a way to delete content from a context menu. `deleteContent` is really just a fallback in case there is no other action covering the action.

So I don't think there is inconsistency there - direction is always collected and specified. Only if there is no direction information we use `deleteContent`. I think this is clear from the description in the column "User's expression of intention".

**Comment by johanneswilm** [comment]
I just checked how it's working on Safari and Chrome without any modifier keys. Both seem to have bugs when deleting selected content and both seem to work correctly when the selection is not collapsed (`deleteContentForward` for the `Delete`-key and `deleteContentBackward` for the `Backspace`-key).

When content is selected, here is what I found:
### Chrome
`Backspace` produces beforeinput: `deleteContentBackward` input: `deleteContentBackward`.
`Delete` produces beforeinput: `deleteContentForward` input **`deleteContentBackward`**.

### Safari
`Backspace` produces beforeinput: `deleteContentBackward` input: `deleteContentBackward`.
`Delete` produces beforeinput: **`deleteContentBackward`** input **`deleteContentBackward`**.

(In bold all the ones that are incorrect. (cc @NavidZ @whsieh))

It seems obvious that they tried to follow the specification, but there are a few bugs in there. Obviously, if the user hits more advanced key code combinations that are not handled for some reason yet content is still deleted, it would seem to make sense to fall back to one of either `deleteContentBackward` or `deleteContentForward` before falling back to `deleteContent`. But maybe those browsers simply say that they don't really know what direction the deletion was going in because the command was not recognized, and that also makes sense to some degree.

**Comment by whsieh** [comment]
Ah, yes, that should emit "deleteContentForward".

I filed <https://bugs.webkit.org/show_bug.cgi?id=191111> to track this in WebKit.

**Comment by masayuki-nakano** [comment]
@johanneswilm 

> I think the wording is quite clear on that.

I don't think so at least Chrome/Safari changes their behavior. I think that spec should mention preferred `inputType` value of different state of selection.

But I'm really happy since you clearly said what should be used when selection is **not** collapsed.

I investigated non-collapsed selection cases:

**Chrome**

`ctrl` + `Backspace` (Win): beforeinput: **`deleteWordBackward`** input: **`deleteWordBackward`**
`option` + `Backspace` (macOS): beforeinput: **`deleteWordBackward`** input: **`deleteWordBackward`**
`command` + `Backspace` (macOS): beforeinput: **`deleteSoftLineBackward`** input: **`deleteSoftLineBackward`**

**Safari**

`option` + `Backspace` (macOS): beforeinput: **`deleteContent`** input: **`deleteContent`**.
`command` + `Backspace` (macOS): beforeinput: **`deleteContent`** input: **`deleteContent`**

All of them should be `deleteContentBackward` under current spec.

@whsieh Could you track this too in the bug?

I'll update https://bugs.chromium.org/p/chromium/issues/detail?id=899624 for Chrome.
(And I'll implement Firefox as so.)

**Comment by whsieh** [comment]
> @whsieh Could you track this too in the bug?

👍

**Comment by Monata** [comment]
On Windows using chrome using backspace returns `deleteContent`,
Using ctrl + backspace returns `deleteWordBackward`,
But after using ctrl + backspace , if delete is pressed it only deletes a letter but still returns `deleteWordBackward`.
